average.js

```diff
@@ -1,6 +1,23 @@
 
 function average(numbers) {
-  return numbers.reduce((p, c)=> p + c, 0) / numbers.length;
+  
+  let result = 0;
+  let countNan = 0;
+  
+  for (let i = 0; i<numbers.length; i++)
+  {
+      
+    if(isNaN(numbers[i]))
+    {
+      countNan = countNan + 1;
+    }
+    else
+    {
+      result = result + numbers[i];
+    }
+  }
+  let newLength = numbers.length - countNan; 
+  return result/newLength;
 }
 
 module.exports = {average};
```

Comment:
Try a new test: The tests in `test/average.test.js` express the functionality of the `average` function.
The functionality accepts a sequence of readings from an IoT sensor. The sensor sends `NaN` when something goes wrong in measurement, which we have ignored.
Problem: If the function sends a lot of NaNs and very few readings, something is wrong. However, the `average` function ignores all the NaNs and still computes the average, as if nothing went wrong.
We need to add functionality to the average function to return `NaN`, when most of the input-numbers are `NaN`.
You need to:

- Reason out what 'most of the numbers' means: E.g., determine the number of NaNs in the input which would indicate that the sensor is unreliable. Mention any references you used to arrive at this logic
- Add a test that supplies such inputs and expects the output to be NaN
- Add code that satisfies this functionality, while not breaking existing functionality.

Reply:
According to me the maximum missing values should be in the threshold of 25-30%, beyond which I used to drop the variable for analysis.
I added a test, to check for inputs more than 30% data missing to be giving the output as NaN.

Reply:
`
function average(numbers) {
    let result = 0;
    let countNan = 0;
    
    numbers.forEach((number) => isNaN(number) ? countNan+=1 : result+=number);
    
    // Sensor Malfunction check!
    if(countNan > Math.ceil(0.30*numbers.length)){
        return NaN;
    }
    else{
        return result / (numbers.length - countNan)
    }
}
module.exports = {average}; 
`

Reply:
average.test.js
`
const {expect} = require('chai');
const {average} = require('../average');

it('computes average of a list of numbers', ()=> {
  // floating point numbers cannot be compared for equality,
  // hence allowing a delta tolerance
  expect(average([1, 2, 3, 4])).to.be.approximately(2.5, 0.01);
});

it('reports the average as NaN on an empty list', ()=> {
  expect(average([])).to.be.NaN;
});

it('ignores NaN in the input', ()=> {
  expect(average([1, NaN, 2])).to.be.approximately(1.5, 0.01);
});

it('Reports the average as NaN, if sensor is malfunctioned', () => {
  // Sensor is malfunctioned when more than 30%,
  // data is missing
  expect(average([1,2,3,4,5,NaN,NaN,NaN,NaN])).to.be.NaN;
});
`